### ISSUE TITLE
How to produce 68 points of face landmark?

### ISSUE THREAD

**Issue by herleeyandi** [CongWeilin/mtcnn-caffe#22]
Hi I am new in face landmark, I think this is really excellent work. This really help me so much, thank you Cong Wei Lin. I have one question for you and all, could you specify or give some suggestions, how can I produce 68 points of landmark?, should I train again? 
-Thank you-

**Comment by CongWeilin** [comment]
you can replace 5 points to 68 points, but i think it will be much more difficult.
1. more points will have more punishment value to back propagate. it will disturb the other two training task.
2.68 points contains 17 contour point but 5 points do not contain. Usually we should separate contour points with inner points because they have different mount of useless(background information).

However, you could try them yourself. Good Luck!

**Comment by herleeyandi** [comment]
Thank you CongWeilin for the response and suggestion, 1 star for you :) . I am still confuse "more points will have more punishment value to back propagate. it will disturb the other two training task". 

1. According to the paper, just 48Net (O-Net) is doing landmarking, so basically I just need to train this network right?
2. It doesn't matter for me for the training process will be more, but if I using 68 points, does it need more time when testing?

Thank you.

**Comment by CongWeilin** [comment]
Thanks.

1. According to  part3 Experiments ,the paper also do landmarking in 12net, 24net. But 12 and 24 is too small for 68 points. So you can just replace 48net from 5 points to 68 points.

2. Training 68 points will have more loss. For example, if you use the same learning rate and same net-work train 5 points and 68 points. The 5 points can converge well, but 68 points might show NAN. 68 points' back propagate value will be quite large and will change convolution layers' params a lot. And this will make training cls or roi task harder... Also, separate training contour points and inner points is also quite important because their background information are different.

3. I suggest that you can train a 68 points localization model first, at that time you will have more experience in 68 points training and replace 5 points with 68 points will be quite easy.

Good luck

**Comment by herleeyandi** [comment]
1) So for doing the same thing, I just try to do it in 48net. In this step do I need 48 net hard example?
2) In which place of the code preparation I can change into 68 points?, or basically it foloow the annotation of our training data?
3) If I doing this, basically I just follow the "step 4" in your readme right?, but I still don't get it `step4. Similar to last step, Run gen_24net_hard_example.py to generate hard sample. Run gen_48net_data.py. Combine these output and generate imdb.` Can you give me more hits, seems I am new in this kind of stuff?
4) What you mean by `I suggest that you can train a 68 points localization model first,, at that time you will have more experience in 68 points training and replace 5 points with 68 points will be quite easy.` Did you means that if I have a model from 68 points in 48 net, it just daraw the point in the demo program?

Thank you so much

**Comment by herleeyandi** [comment]
It also failed to `import tools` here is my code. Did you means this is tools inside caffe folder?
`import sys`
`sys.path.append('.')`
`sys.path.append('/home/mspl/caffe/python')`
`import tools`
`import caffe`
`import cv2`
`import numpy as np`
`import os`
`from utils import *`

**Comment by goodloop** [comment]
@herleeyandi try change "cls_score" to "prob1"

**Comment by herleeyandi** [comment]
@goodloop Thank you so much, have you able to produce pts.lmdb?, what kind of dataset did you use?
-Thank you-

**Comment by goodloop** [comment]
no pts.imdb, maybe pts is not needed in 12net.  
I use  Widerface.
Btw, too many small files generated cause toomuch time preparing data.

**Comment by wolf345** [comment]
@herleeyandi @goodloop Did you solve this problem?--no pts.imdb
